src/opnmidi_midiplay.cpp

```diff
@@ -208,8 +212,33 @@ void OPNMIDIplay::resetMIDI()
 void OPNMIDIplay::TickIterators(double s)
 {
     Synth &synth = *m_synth;
-    for(uint16_t c = 0; c < synth.m_numChannels; ++c)
-        m_chipChannels[c].addAge(static_cast<int64_t>(s * 1e6));
+    for(uint32_t c = 0, n = synth.m_numChannels; c < n; ++c)
+    {
+        OpnChannel &ch = m_chipChannels[c];
+        ch.addAge(static_cast<int64_t>(s * 1e6));
+    }
+
+    // Resolve "hell of all times" of too short drum notes
+    for(size_t c = 0, n = m_midiChannels.size(); c < n; ++c)
```

Comment:
I think, instead of this loop, make the array of "delayed note offs", e.g. register all delayed notes in it, and iterate it instead of making extra fetch of all notes of all channels. :thinking: :fox_face:

Reply:
Yep, this is the small thing which I've been on the fence about.
but are there ever so many notes that this is going to make a difference? so I took the solution of simplicity

Reply:
I mean the fact of fetching of all channels and all notes are even not needing for an extra checking. My idea was forwarding any note-offs are addressed to too short notes. I.e. instead of execution, put them into an array which will be checked each loop (e.g. put a note on watching), and when this note reaches specified length, execute the note off for this and drop the array entry.

Reply:
Yes, I understand this exactly. This is what I mean to say:
I can't say supposed performance gain is going to be or not significant, I suppose it's not necessarily so important based on points:

1. it's infrequent operation: once per cycle, having repetition interval in the milliseconds
2. activenotes list is already minimized by itself, it's a talk of adding more minimization on what is current, for this case in particular
3. memorizing more data, it means also more chance of cache miss, and also code complexity is up. (because state must keep in sync)

Reply:
> memorizing more data, it means also more chance of cache miss, and also code complexity is up. (because state must keep in sync)

Only "note off"s, but yeah, just now I have got an another though what would happen:
- two notes are near each other, but they are both too short
- first note off
- second note on (note off of previous still not happen)
- finally happens first note off, but, the second note is off which have already overrode first [until the ability to play multiple notes on the same key will be implemented]
So, your scheme is saving from a sort of like this.

Can you add the flag "has delayed notes" like was made on Portamento? to don't loop this when is no waiting notes.

Reply:
Yeah I forgot about doing this, let's add it.